### PR TITLE
moreh_mean_backward: register output_grad as BufferBinding to fix stale-address cache hits

### LINKED ISSUE #44730: [CI Failure] tt-train-cpp-unit-tests (blackhole, P150b) - AdamWCompositeFullTest.AdamWCompositeTest
### Pipeline Name

tt-train-cpp-unit-tests (blackhole, P150b)

### Test Name

AdamWCompositeFullTest.AdamWCompositeTest

### Path to Test in Repo

tt-train/tests/optimizers/adamw_composite_test.cpp

### Non-Deterministic (ND) failure?

No

### Failure type

Other

### Link to Failing Pipeline for non-ND or Custom Dispatch Run for ND

https://github.com/tenstorrent/tt-metal/actions/runs/25741886390/job/75599262997

### Failure Signature Snippet

```shell
Note: Google Test filter = AdamWCompositeFullTest.AdamWCompositeTest
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from AdamWCompositeFullTest
[ RUN      ] AdamWCompositeFullTest.AdamWCompositeTest
2026-05-12 15:19:45.577 | info     |             UMD | Creating TopologyDiscovery for architecture: blackhole (topology_discovery.cpp:73)
2026-05-12 15:19:45.581 | warning  |             UMD | Low power mode is disabled while UMD holds open file descriptors. The device will return to low power mode once all file descriptors are closed. (topology_discovery.cpp:141)
2026-05-12 15:19:45.582 | info     |             UMD | Established firmware bundle version: 19.8.0 (topology_discovery.cpp:537)
2026-05-12 15:19:45.582 | info     |             UMD | Firmware bundle version 19.8.0 on the system is newer than the latest fully tested version 19.5.0 for blackhole architecture. Newest features may not be supported. (topology_discovery.cpp:569)
2026-05-12 15:19:45.584 | info     |          Device | Opening user mode device driver (tt_cluster.cpp:225)
2026-05-12 15:19:45.584 | info     |             UMD | Creating TopologyDiscovery for architecture: blackhole (topology_discovery.cpp:73)
2026-05-12 15:19:45.587 | warning  |             UMD | Low power mode is disabled while UMD holds open file descriptors. The device will return to low power mode once all file descriptors are closed. (topology_discovery.cpp:141)
2026-05-12 15:19:45.588 | info     |             UMD | Established firmware bundle version: 19.8.0 (topology_discovery.cpp:537)
2026-05-12 15:19:45.588 | info     |             UMD | Firmware bundle version 19.8.0 on the system is newer than the latest fully tested version 19.5.0 for blackhole architecture. Newest features may not be supported. (topology_discovery.cpp:569)
2026-05-12 15:19:45.589 | info     |             UMD | Harvesting masks for Chip 0: Tensix: 0xc0 DRAM: 0x0 ETH: 0x120 PCIe: 0x2 L2CPU: 0x0 (cluster.cpp:270)
2026-05-12 15:19:45.589 | info     |             UMD | Initializing iommu for sysmem (size: 0x40000000). (silicon_sysmem_manager.cpp:337)
2026-05-12 15:19:45.589 | info     |             UMD | Allocating sysmem for IOMMU (size: 0x40000000). (silicon_sysmem_manager.cpp:343)
2026-05-12 15:19:45.955 | info     |             UMD | Opening local chip ids/PCIe ids: {0}/[0] and remote chip ids {} (cluster.cpp:159)
2026-05-12 15:19:45.956 | info     |             UMD | IOMMU: enabled (cluster.cpp:133)
2026-05-12 15:19:45.956 | info     |             UMD | KMD version: 2.7.0 (cluster.cpp:136)
2026-05-12 15:19:45.957 | info     |             UMD | Starting devices in cluster (cluster.cpp:1042)
2026-05-12 15:19:46.790 | info     |             UMD | Mapped sysmem without hugepages to IOVA 0x40000000; NOC address 0x1000000000000000 (silicon_sysmem_manager.cpp:394)
2026-05-12 15:19:47.025 | info     |     Distributed | Using auto discovery to generate mesh graph. (metal_env.cpp:332)
2026-05-12 15:19:47.025 | info     |     Distributed | Constructing control plane using auto-discovery (no mesh graph descriptor). (metal_env.cpp:394)
2026-05-12 15:19:47.025 | info     |             UMD | Creating TopologyDiscovery for architecture: blackhole (topology_discovery.cpp:73)
2026-05-12 15:19:47.028 | warning  |             UMD | Low power mode is disabled while UMD holds open file descriptors. The device will return to low power mode once all file descriptors are closed. (topology_discovery.cpp:141)
2026-05-12 15:19:47.029 | info     |             UMD | Established firmware bundle version: 19.8.0 (topology_discovery.cpp:537)
2026-05-12 15:19:47.029 | info     |             UMD | Firmware bundle version 19.8.0 on the system is newer than the latest fully tested version 19.5.0 for blackhole architecture. Newest features may not be supported. (topology_discovery.cpp:569)
2026-05-12 15:19:47.033 | warning  |          Always | Unknown motherboard '' for chip_id=0 (bus_id=0x1) — defaulting tray_id to 0. Add this motherboard and its bus IDs to mobo_to_bus_ids in physical_system_discovery.cpp. (physical_system_discovery.cpp:62)
2026-05-12 15:19:47.033 | info     |          Fabric | TopologyMapper: Logical Multi-Mesh Adjacency Map: (topology_mapper.cpp:1474)
2026-05-12 15:19:47.033 | info     |          Fabric | 
=== Logical Mesh-Level Graph Adjacency Map ===
Total nodes: 1
Degree histogram: {0:1}
 (topology_solver.tpp:94)
2026-05-12 15:19:47.033 | info     |          Fabric | 
=== Logical Mesh 0 Internal Graph Adjacency Map ===
Total nodes: 1
Degree histogram: {0:1}
 (topology_solver.tpp:94)
2026-05-12 15:19:47.033 | info     |          Fabric | TopologyMapper: Physical Multi-Mesh Adjacency Map: (topology_mapper.cpp:1485)
2026-05-12 15:19:47.033 | info     |          Fabric | 
=== Physical Mesh-Level Graph Adjacency Map ===
Total nodes: 1
Degree histogram: {0:1}
 (topology_solver.tpp:94)
2026-05-12 15:19:47.033 | info     |          Fabric | 
=== Physical Mesh 0 Internal Graph Adjacency Map ===
Total nodes: 1
Degree histogram: {0:1}
 (topology_solver.tpp:94)
2026-05-12 15:19:47.033 | info     |          Fabric | Starting multi-mesh mapping: 1 logical mesh(es) to 1 physical mesh(es) (topology_mapper_utils.cpp:1469)
2026-05-12 15:19:47.033 | info     |          Fabric | Multi-mesh mapping attempt 1/2: Trying inter-mesh mapping (topology_mapper_utils.cpp:1510)
2026-05-12 15:19:47.034 | info     |          Fabric | Attempt 1: Inter-mesh mapping succeeded, found 1 mesh pair(s) (topology_mapper_utils.cpp:1539)
2026-05-12 15:19:47.034 | info     |          Fabric | Attempt 1: Mapping 1 mesh pair(s) (topology_mapper_utils.cpp:1547)
2026-05-12 15:19:47.034 | info     |          Fabric | Attempt 1: Mapping mesh 0 -> 0 (topology_mapper_utils.cpp:1554)
2026-05-12 15:19:47.034 | info     |          Fabric | Multi-mesh mapping succeeded after 1 attempt(s): 1 mesh pair(s) mapped (topology_mapper_utils.cpp:1686)
2026-05-12 15:19:47.048 | info     |    BuildKernels | Using pre-compiled firmware from: /work/tt_metal/pre-compiled/2010209910287226822/ (build_env_manager.cpp:260)
2026-05-12 15:19:47.738 | info     |           Metal | [Real-time profiler] Device 0 sync complete: 100 samples, frequency=1.349908 GHz, device_time_at_sync=479371602972 cycles (realtime_profiler_manager.cpp:1077)
/work/tt-train/tests/optimizers/adamw_composite_test.cpp:76: Failure
Expected: (losses.back()) < (1e-3F), actual: 0.0224609375 vs 0.001
[  FAILED  ] AdamWCompositeFullTest.AdamWCompositeTest (8413 ms)
[----------] 1 test from AdamWCompositeFullTest (8413 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (8413 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] AdamWCompositeFullTest.AdamWCompositeTest

 1 FAILED TEST
```

### Additional info

```shell
Please investigate changes made in #42481 to moreh_mean_backward operation. After revert test passes.
```

### PR DESCRIPTION
**Pull request by dgomezTT**
## Summary

Fixes #44730 — `tt-train` `AdamWCompositeFullTest.AdamWCompositeTest` failing on Blackhole P150b after my migration in #42481.

## Root cause

In the post-migration factory, the **writer** kernel's `input_grad` was correctly converted to a `BufferBinding` via `emplace_runtime_args(core, {input_grad.buffer(), ...})`. The **reader** kernel still wrote `output_grad.buffer()->address()` as a raw `uint32_t` into its runtime arg list (because of its variable-length arg layout), with no corresponding `BufferBinding`.

The descriptor framework only re-patches addresses that are registered as `BufferBindings` on cache hits. So on every cache-hit invocation, the reader kept reading from whatever address was set on the first call, regardless of what `output_grad` the caller passed in.

## Why CI didn't catch it pre-merge

Same reason Pavle's same-class SDPA bug (8635db0bafe / #45121) was caught only by a downstream consumer test:
- Cache-miss path uses the address directly → correct on first call.
- Single-shot unit tests never exercise cache-hit, so they pass.
- The device allocator typically reuses the same slot for a freshly-allocated gradient tensor on N150, so even the cache-hit tests happen to land on the right address.
- Blackhole P150b's allocation pattern in the AdamW training loop *does* hand out a fresh address each step, so the reader points at stale memory → garbage gradient → AdamW fails to converge: `losses.back() = 0.0224 vs < 1e-3`.

## The fix

Switch the reader to `KernelDescriptor::RTArgList` and pass `output_grad.buffer()` (Buffer\*) instead of `.buffer()->address()` (uint32_t). The framework auto-registers the BufferBinding at the right arg position and re-patches the address on every cache hit. Variable-length dim vectors remain raw uint32 args.

## Test plan

- [x] Full `tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_mean.py` — 76 passed, 72 skipped on N150 (no regression).
- [x] `test_moreh_mean_backward_callback` (2-iteration cache-hit test) passes — confirms the new arg-list shape is correct; it doesn't independently reproduce the BH allocator pattern.
- [ ] tt-train `AdamWCompositeFullTest.AdamWCompositeTest` on Blackhole P150b — needs the BH lab to confirm convergence is restored. This is what nightly will catch.

## Audit follow-up

`moreh_sum_backward_program_factory.cpp:250` has the exact same pattern (reader's `output_grad.buffer()->address()` raw-pushed in a `create_descriptor`-based factory with no override callback). Not fixed in this PR per scope. If a downstream test starts failing for sum_backward in a training loop, that's where to look.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:dgomez/fix-moreh-mean-backward-buffer-binding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:dgomez/fix-moreh-mean-backward-buffer-binding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:dgomez/fix-moreh-mean-backward-buffer-binding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:dgomez/fix-moreh-mean-backward-buffer-binding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:dgomez/fix-moreh-mean-backward-buffer-binding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:dgomez/fix-moreh-mean-backward-buffer-binding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:dgomez/fix-moreh-mean-backward-buffer-binding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:dgomez/fix-moreh-mean-backward-buffer-binding)

### CODE REVIEW

**Review by zbaczewskiTT — APPROVED** [review]
Thank you!

Verified on P100a the loss is identical on this branch and parent of #42481

**Review by bbradelTT — APPROVED** [review]
